### PR TITLE
encryption: fix unexpected error code for wrong master-key(tikv#17410)

### LINKED ISSUE #17410: encryption: error occurs when rotating master key
## Bug Report
```
[2024/08/20 11:16:19.933 +00:00] [FATAL] [lib.rs:478] ["Encryption failed to initialize: Cloud KMS error Key error: Kms error OAuth(HttpStatus(400), \"https://cloudkms.googleapis.com/v1/projects/qa-infra-dev/locations/us-west1/keyRings/tikv-kms-testing/cryptoKeys/tikv_kms_testing_v2/:decrypt?alt=json\") decrypt encrypted key failed. code: KV:Cloud:Unknown"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}
             at /workspace/source/tikv/components/tikv_util/src/lib.rs:477:18
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2029:9
      std::panicking::rust_panic_with_hook
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:783:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:657:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:171:18
   4: rust_begin_unwind
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:645:5
   5: core::panicking::panic_fmt
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:72:14
   6: server::common::TikvServerCore::init_encryption::{{closure}}
             at /workspace/source/tikv/components/server/src/common.rs:277:13
      core::result::Result<T,E>::map_err
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:829:27
      server::common::TikvServerCore::init_encryption
             at /workspace/source/tikv/components/server/src/common.rs:272:39
   7: server::server::run_impl
             at /workspace/source/tikv/components/server/src/server.rs:165:5
   8: server::server::run_tikv
             at /workspace/source/tikv/components/server/src/server.rs:227:5
      tikv_server::main
             at /workspace/source/tikv/cmd/tikv-server/src/main.rs:249:31
   9: core::ops::function::FnOnce::call_once
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /root/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:155:18
  10: main
  11: <unknown>
  12: __libc_start_main
  13: <unknown>
```

### What version of TiKV are you using?
v8.2.0.

### What operating system and CPU are you using?
independent of OS and CPU.

### Steps to reproduce
1. Enable GCP KMS encryption.
2. Rotate to a new master key.
3. Reload TiKV.

### What did you expect?
TiKV should reload success without errors.

### What did happened?
TiKV failed to reload, and errors are in the TiKV log.

### PR DESCRIPTION
**Pull request by hhwyt**
### What is changed and how it works?

Issue Number: close #17410 

What's Changed:

```commit-message
This PR fixes a bug encountered in the GCP KMS master-key rotation
scenario. When decryption fails using the current master-key, the
`KsmBackend` is expected to return a `WrongMasterKey` error rather
than any other error type. The higher-level caller will only retry
with the previous master-key if a `WrongMasterKey` error is received.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - Tested using the reproduction steps from #17410, and everything is working correctly.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Fix a bug that cannot rotate encryption master key when master key is stored in KMS
```

### CODE REVIEW

**Review by LykxSassinator — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

File: components/encryption/src/master_key/kms.rs [hunk 1]
```diff
@@ -241,21 +262,36 @@ mod tests {
...
+        KmsBackend::new(Box::new(FakeKms::new(
+            plainkey,
+            should_decrypt_data_key_fail,
+        )))
+        .unwrap()
+    }
+
     #[test]
     fn test_kms_backend() {
-        // See more http://csrc.nist.gov/groups/STM/cavp/documents/mac/gcmtestvectors.zip
```
Comment (overvenus) [inline components/encryption/src/master_key/kms.rs:246]: We should keep this comment, that's where keys come from.

Comment (hhwyt) [reply components/encryption/src/master_key/kms.rs:246]: This was accidentally deleted. I'll submit a PR to bring it back in next weak.

**Comment by Tristan1900** [comment]
out of curiosity since I'm also using this part of the code, why would kms fail to decrypt after auto rotating master key? Mostly likely I'm missing something, but my understanding is that after rotation the key id is still the same and only the key version changed, so if you try to decrypt the old version kms will automatically figure out and do the decryption. In theory it should not have any problems but maybe this is a different type of master key rotation? Thanks!

**Comment by hhwyt** [comment]
> @hhwyt out of curiosity since I'm also using this part of the code, why would kms fail to decrypt after auto rotating master key? Mostly likely I'm missing something, but my understanding is that after rotation the key id is still the same and only the key version changed, so if you try to decrypt the old version kms will automatically figure out and do the decryption. In theory it should not have any problems but maybe this is a different type of master key rotation? Thanks!

@Tristan1900 KMS will fail if we rotate to a different master key that isn't just a new version of the previous master key.

**Comment by Tristan1900** [comment]
@hhwyt thanks for the response. ok that makes sense if we are manually rotating to a different master key with different key id.
If that's the case maybe we could add some checks at [here](https://github.com/tikv/tikv/pull/17413/files#diff-0f18341af3dc671328d8f26275bacc9676378054ce65b6eba9e561d69511b4c9R103) to not only check the vendor but also key id? Feels like that's more explicitly handing this case. Right now we are considering any failures from the kms to be wrong master key error, I'm not sure if that's true given it could be network error or kms unavailable etc. 
let me know your thoughts! thanks!

**Comment by hhwyt** [comment]
> @hhwyt thanks for the response. ok that makes sense if we are manually rotating to a different master key with different key id. If that's the case maybe we could add some checks at [here](https://github.com/tikv/tikv/pull/17413/files#diff-0f18341af3dc671328d8f26275bacc9676378054ce65b6eba9e561d69511b4c9R103) to not only check the vendor but also key id? Feels like that's more explicitly handing this case. Right now we are considering any failures from the kms to be wrong master key error, I'm not sure if that's true given it could be network error or kms unavailable etc. let me know your thoughts! thanks!

Yes, you are right. The errors definitely need to be further categorized. I noticed this issue as well, which is why I pointed out in [this comment](https://github.com/tikv/tikv/pull/17419#issuecomment-2315410951). I'll fixed it soon. Thanks for you feedback!

**Comment by hhwyt** [comment]
@Tristan1900 It seems like `classify_decrypt_error`  could help us identify the wrong master key case.

**Comment by hhwyt** [comment]
> > @hhwyt thanks for the response. ok that makes sense if we are manually rotating to a different master key with different key id. If that's the case maybe we could add some checks at [here](https://github.com/tikv/tikv/pull/17413/files#diff-0f18341af3dc671328d8f26275bacc9676378054ce65b6eba9e561d69511b4c9R103) to not only check the vendor but also key id? Feels like that's more explicitly handing this case. Right now we are considering any failures from the kms to be wrong master key error, I'm not sure if that's true given it could be network error or kms unavailable etc. let me know your thoughts! thanks!
> 
> Yes, you are right. The errors definitely need to be further categorized. I noticed this issue as well, which is why I pointed out in [this comment](https://github.com/tikv/tikv/pull/17419#issuecomment-2315410951). I'll fixed it soon. Thanks for your feedback!

This issue has been logged in https://github.com/tikv/tikv/issues/17457.

**Pull request merged**